### PR TITLE
meta: updated collaborator guide and content-vs-code

### PR DESCRIPTION
**Pull request by ovflowd**
## Description

This PR adds relevant technical information to our Collaborators and updates the now-outdated CONTENT_VS_CODE file.

### CODE REVIEW

**Comment by ovflowd** [comment]
cc @tniessen @nodejs/website for review and awareness.

File: COLLABORATOR_GUIDE.md [hunk 1]
```diff
@@ -15,79 +15,47 @@
...
-lack of consensus may indicate the need to elevate the discussion to the
-Website Group for resolution (see below).
-
-All bug fixes require a test case that demonstrates the defect. The
-test should _fail_ before the change, and _pass_ after the change.
-
-All pull requests that modify executable code should be subjected to
-continuous integration tests on the
-[project CI server](https://ci.nodejs.org/).
+We recommend Collaboratory to follow the guidelines on the [Contributing Guide](./CONTRIBUTING.md#before-merging) for reviewing and merging Pull Requests.
```
Comment (mikeesto) [inline COLLABORATOR_GUIDE.md]:
```suggestion
We recommend collaborators follow the guidelines on the [Contributing Guide](./CONTRIBUTING.md#before-merging) for reviewing and merging Pull Requests.
```

File: COLLABORATOR_GUIDE.md [hunk 2]
```diff
@@ -279,6 +245,151 @@ export default { component: NameOfComponent } as Meta;
 - We recommend reading previous Storybooks from the codebase for inspiration and code guidelines.
 - If you need to decorate/wrap your Component/Story with a Container/Provider, please use [Storybook Decorators](https://storybook.js.org/docs/react/writing-stories/decorators)
 
+## Remarks on Technologies Used
+
+The Node.js Website is a somewhat complex Application. And hecen sometimes non-trivial solutions were adopted to solve certain technical challenges.
```
Comment (mikeesto) [inline COLLABORATOR_GUIDE.md]:
```suggestion
The Node.js Website is a somewhat complex application and at times non-trivial solutions have been implemented to solve certain technical challenges.
```

File: COLLABORATOR_GUIDE.md [hunk 3]
```diff
@@ -279,6 +245,151 @@ export default { component: NameOfComponent } as Meta;
...
+Past Issues, Conversations on Slack and GitHub Discussions are the go-to-go for most historical purposes, but we've found important to highlight some of these decisions here.
```
Comment (mikeesto) [inline COLLABORATOR_GUIDE.md]:
```suggestion
Historical decision making can be largely found on past issues, conversations on Slack and GitHub discussions. However, we also wish to highlight some of the notable development decisions that we have made here.
```

File: COLLABORATOR_GUIDE.md [hunk 4]
```diff
@@ -279,6 +245,151 @@ export default { component: NameOfComponent } as Meta;
...
+
+### Why Next.js?
+
+We've found that Next.js is simply versatile, hackable, stable, community-maintained and has a great ecosystem.
+The reasoning goes deeper, but as a long-term Framework it is the most suitable choice.
+
+#### Why we have Static Exports?
```
Comment (mikeesto) [inline COLLABORATOR_GUIDE.md]:
```suggestion
#### Why do we continue to support static builds?
```

File: COLLABORATOR_GUIDE.md [hunk 5]
```diff
@@ -279,6 +245,151 @@ export default { component: NameOfComponent } as Meta;
...
+#### What is `next.dynamic.mjs`?
+
+Our whole Website uses a custom renderer for rendering the pages.
+As you might have seen, within the `pages` directory we have [Next.js Dynamic Route](https://nextjs.org/docs/pages/building-your-application/routing/dynamic-routes) named `[...path].tsx` that matches against all possible routes of the Website.
+
+This means that each `.md(x)` file within `pages/` is not rendered by Next.js regular App Tree (`pages/_document.tsx` and `pages/_app.tsx`) but a custom render tree.
+
+This custom render uses `getStaticPaths` and [Incremental Static Generation](https://nextjs.org/docs/pages/building-your-application/data-fetching/incremental-static-regeneration) to generate the full list of supported pages of the Website.
+For example, this allows us to generate Localized Pages for every page that is not translated, by telling Next.js to create a localised path.
+`next.dynamic.mjs` is responsible for getting a full list of the source pages (`pages/en`) and forcefully get a diff of what is translated and whatnot.
```
Comment (mikeesto) [inline COLLABORATOR_GUIDE.md]:
```suggestion
`next.dynamic.mjs` is responsible for getting a full list of the source pages (`pages/en`) and identifying which pages have been translated.
```

File: COLLABORATOR_GUIDE.md [hunk 6]
```diff
@@ -279,6 +245,151 @@ export default { component: NameOfComponent } as Meta;
...
+
+Non-translated pages will have their Localized contexts and translated React message-bags (`react-intl`) but the content will be the same as the source page (English).
+Whereas localized pages will have localized context and content.
+
+This custom solution is also able to decide what paths should be compiled during runtime and whatnot.
```
Comment (mikeesto) [inline COLLABORATOR_GUIDE.md]:
```suggestion
This custom solution is also able to decide what paths should be compiled during runtime.
```

File: COLLABORATOR_GUIDE.md [hunk 7]
```diff
@@ -279,6 +245,151 @@ export default { component: NameOfComponent } as Meta;
...
+This is a combination of rules defined on `next.constants.mjs` and `[...path].tsx`.
+
+Finally, `[...path].tsx` uses the root `theme.tsx` as layout.
```
Comment (mikeesto) [inline COLLABORATOR_GUIDE.md]:
```suggestion
The `[...path].tsx` file ultimately utilizes the `theme.tsx` file as its layout source.
```

File: COLLABORATOR_GUIDE.md [hunk 8]
```diff
@@ -279,6 +245,151 @@ export default { component: NameOfComponent } as Meta;
...
+This finally loads the Layout Provider and MDX Provider and then any children content/Component from the Layout Provider + the transformed MDX content from the `.md(x)` source page.
```
Comment (mikeesto) [inline COLLABORATOR_GUIDE.md]:
```suggestion
This setup enables the loading of the Layout Provider and MDX Provider, which in turn, encapsulate and manage the rendering of any child content or components. This includes both content and components provided by the Layout Provider and the transformed MDX content originating from the `.md(x)` source page.
```

File: COLLABORATOR_GUIDE.md [hunk 9]
```diff
@@ -279,6 +245,151 @@ export default { component: NameOfComponent } as Meta;
...
+#### What is `next.data.mjs`?
+
+This file is responsible for loading, fetching and generating build-time required information such as Node.js Release data, Blog Posts Metadata (for pagination and indexation), RSS feeds, etc.
+
+#### What is `site.json`?
+
+This file is used for defining Website Metadata, such as which RSS feeds should be generated, Social Media Information and other Metadata used during the Website build-time.
+We use a JSON format to ease collaboration.
+
+#### What is `next.locales.mjs` and why not Next.js built-in i18n?
```
Comment (mikeesto) [inline COLLABORATOR_GUIDE.md]:
```suggestion
#### What is `next.locales.mjs` and why not use Next.js built-in i18n?
```

File: COLLABORATOR_GUIDE.md [hunk 10]
```diff
@@ -279,6 +245,151 @@ export default { component: NameOfComponent } as Meta;
...
+- Redirects change the original URL to the new one. The URLs do not necessarily need to be external, but they in general should be.
+
+This file contains a simple template engine that supports `/:locale` to indicate that this path should also be available under all available locales as prefix.
+
+#### Our Next.js Middleware
+
+We have a simple Next.js Middleware that is responsible for handling initial Locale detection and redirection.
+It detects browser locales and redirects to the most suitable locale for the user. And it fallbacks to `/en` if no suitable locale is found.
+
+#### How Layouts Work within this Repository?
```
Comment (mikeesto) [inline COLLABORATOR_GUIDE.md]:
```suggestion
#### How do layouts work?
```

File: COLLABORATOR_GUIDE.md [hunk 11]
```diff
@@ -279,6 +245,151 @@ export default { component: NameOfComponent } as Meta;
...
+Each Page layout is decided within their Markdown's Frontmatter as `layout: name-of-layout.hbs`.
+
+### Why MDX?
+
+MDX is a great format for writing content. It allows us to write Markdown and React Components in the same file.
+It also allows us to use React Components within Markdown files.
+This is a great advantage for us, as we can use React Components to create complex content structures and layouts.
+It also supports the `unified` ecosystem, which allows us to extend the MDX functionality with ease.
+
+Some of the Plugins used here include, but not limited to:
```
Comment (mikeesto) [inline COLLABORATOR_GUIDE.md]:
```suggestion
Some of the plugins that we use include:
```

File: COLLABORATOR_GUIDE.md [hunk 12]
```diff
@@ -279,6 +245,151 @@ export default { component: NameOfComponent } as Meta;
...
+We use `next/fonts` Open Sans as the default font for the Node.js Website.
+The font is configured as a CSS variable and then configured on `tailwind.config.js` as the default font for the Website.
+
+#### Sass Support / IDE Support
+
+We recommend the usage of [Tailwind's Official VS Code Extension](https://marketplace.visualstudio.com/items?itemName=bradlc.vscode-tailwindcss) for IntelliSense Support. You might need to set `.scss` files to use Tailwind as the Interpreter.
+
+### Vercel
+
+We use Vercel as our hosting provider. It is a great platform that allows us to deploy our Website with ease.
```
Comment (mikeesto) [inline COLLABORATOR_GUIDE.md]:
```suggestion
We use Vercel as our hosting provider. It is a great platform that offers an excellent CI/CD pipeline which allows us to deploy our website with ease.
```

File: COLLABORATOR_GUIDE.md [hunk 13]
```diff
@@ -279,6 +245,151 @@ export default { component: NameOfComponent } as Meta;
...
+It also allows us to have a great CI/CD Pipeline that allows us to deploy our Website with ease.
```
Comment (mikeesto) [inline COLLABORATOR_GUIDE.md]:
```suggestion

```

File: COLLABORATOR_GUIDE.md [hunk 14]
```diff
@@ -279,6 +245,151 @@ export default { component: NameOfComponent } as Meta;
...
+- Vercel Deployments are triggered for all other branches during `push` activities.
+- We have a custom install script that executes `npm ci --omit=dev` (the same way we do on our CI Pipelines)
+  - Hence if Builds fail unexpectedly, make sure that your dependency that is being used during build-time is on `dependencies` and not `devDependencies`. Checkout out [DEPENDENCY_PINNING.md](./DEPENDENCY_PINNING.md) for more information.
+- Our sponsorship with Vercel is maintained by the OpenJS Foundation
+
+### Unfamiliar with certain decision of Processes?
+
+A lot of the current structure is due to retro-compatibility, keeping a simple and familiar file structure and keeping files that have historical reasons or needs.
+
+If you're unfamiliar our curious about something, we recommend opening a Discussion on this GitHub Repository.
```
Comment (mikeesto) [inline COLLABORATOR_GUIDE.md]:
```suggestion
If you're unfamiliar or curious about something, we recommend opening a Discussion on this GitHub Repository.
```

File: COLLABORATOR_GUIDE.md [hunk 15]
```diff
@@ -279,6 +245,151 @@ export default { component: NameOfComponent } as Meta;
...
+
+It is important to mention that there are some rules on our Vercel Deployments such as:
+
+- Branches starting with `dependabot` (Dependabot Automated PRs) or `gh` (GitHub Merge Queues) are not deployed to Vercel.
+- Vercel Deployments are triggered for all other branches during `push` activities.
+- We have a custom install script that executes `npm ci --omit=dev` (the same way we do on our CI Pipelines)
+  - Hence if Builds fail unexpectedly, make sure that your dependency that is being used during build-time is on `dependencies` and not `devDependencies`. Checkout out [DEPENDENCY_PINNING.md](./DEPENDENCY_PINNING.md) for more information.
+- Our sponsorship with Vercel is maintained by the OpenJS Foundation
+
+### Unfamiliar with certain decision of Processes?
```
Comment (mikeesto) [inline COLLABORATOR_GUIDE.md]:
```suggestion
### Seeking additional clarification
```

**Review by mikeesto — APPROVED** [review]
LGTM. I've made a couple of suggestions. Feel free to take or leave them as you see fit

**Comment by ovflowd** [comment]
Thank you so much for your suggestions, @mikeesto !!

**Review by shanpriyan — APPROVED** [review]
LGTM!

File: COLLABORATOR_GUIDE.md [hunk 16]
```diff
@@ -279,6 +245,150 @@ export default { component: NameOfComponent } as Meta;
...
+`next.dynamic.mjs` is responsible for getting a full list of the source pages (`pages/en`) and identifying which pages have been translated.
+
+Non-translated pages will have their Localized contexts and translated React message-bags (`react-intl`) but the content will be the same as the source page (English).
+Whereas localized pages will have localized context and content.
+
+This custom solution is also able to decide what paths should be compiled during runtime.
+This is a combination of rules defined on `next.constants.mjs` and `[...path].tsx`.
+
+The `[...path].tsx` file ultimately utilizes the `theme.tsx` file as its layout source.
+This setup enables the loading of the Layout Provider and MDX Provider, which in turn, encapsulate and manage the rendering of any child content or components. This includes both content and components provided by the Layout Provider and the transformed MDX content originating from the `.md(x)` source page.
```
Comment (rodion-arr) [inline COLLABORATOR_GUIDE.md:283]: Looks like we need some more `s` here:

```suggestion
This setup enables the loading of the Layout Provider and MDX Provider, which in turn, encapsulates and manages the rendering of any child content or components. This includes both content and components provided by the Layout Provider and the transformed MDX content originating from the `.md(x)` source page.
```

File: COLLABORATOR_GUIDE.md [hunk 17]
```diff
@@ -279,6 +245,150 @@ export default { component: NameOfComponent } as Meta;
...
+This file is responsible for loading, fetching and generating build-time required information such as Node.js Release data, Blog Posts Metadata (for pagination and indexation), RSS feeds, etc.
+
+#### What is `site.json`?
+
+This file is used for defining Website Metadata, such as which RSS feeds should be generated, Social Media Information and other Metadata used during the Website build-time.
+We use a JSON format to ease collaboration.
+
+#### What is `next.locales.mjs` and why not use Next.js built-in i18n?
+
+Next.js built-in i18n is great, but it is not flexible enough for our needs. We need to be able to generate a full list of supported locales, and we need to be able to generate a full list of supported pages for each locale. We also need to be able to generate a full list of supported pages for each locale, and we need to be able to generate a full list of supported pages for each locale.
```
Comment (rodion-arr) [inline COLLABORATOR_GUIDE.md]: Looks like we have `full list of supported pages for each locale` 3 times?

File: COLLABORATOR_GUIDE.md [hunk 18]
```diff
@@ -279,6 +245,150 @@ export default { component: NameOfComponent } as Meta;
...
+
+#### What is `next.rewrites.mjs`?
+
+This file is responsible for defining the rewrite rules for the Website.
+It is used for defining redirects and other rewrite rules. (Such as Internal Redirects and External ones).
+
+The actual rewrites are defined within `redirects.json`. The file format is simple and separate external redirects with internal rewrites.
+
+- Rewrites do not change the original URL but render the content of the rewritten URL. The reritten path must be valid and exist on the Website.
+- Redirects change the original URL to the new one. The URLs do not necessarily need to be external, but they in general should be.
```
Comment (rodion-arr) [inline COLLABORATOR_GUIDE.md]: The last phrase is not clear to me: 

> but they in general should be.

should be what? Could you please elaborate

File: COLLABORATOR_GUIDE.md [hunk 19]
```diff
@@ -279,6 +245,150 @@ export default { component: NameOfComponent } as Meta;
...
+
+Layouts are defined within the `layouts` folder.
+They are React Components that receive the `children` prop, which is the transformed MDX content of the Markdown file.
+
+Each Page layout is decided within their Markdown's Frontmatter as `layout: name-of-layout.hbs`.
+
+### Why MDX?
+
+MDX is a great format for writing content. It allows us to write Markdown and React Components in the same file.
+It also allows us to use React Components within Markdown files.
```
Comment (rodion-arr) [inline COLLABORATOR_GUIDE.md]: This point is already cleared in the previous sentence, looks like duplication

**Review by rodion-arr — APPROVED** [review]

File: COLLABORATOR_GUIDE.md (hunk 17 above)
Comment (ovflowd) [reply COLLABORATOR_GUIDE.md]: Lol, didn't even notice that!

File: COLLABORATOR_GUIDE.md (hunk 18 above)
Comment (ovflowd) [reply COLLABORATOR_GUIDE.md]: "The URLs do not necessarily need to be external, but they in general should be.", meaning the redirect URLs should generally be external ones. It's implicit context 🤔

Comment (rodion-arr) [reply COLLABORATOR_GUIDE.md]: After re-reading this phrase several times I think I got it now, probably worth rephrasing e.g. `In general the redirect URLs should be external but it is not necessary`

Comment (ovflowd) [reply COLLABORATOR_GUIDE.md]: I think both have the same meaning. At least for me it's easy to understand. Wdyt @mikeesto

Comment (mikeesto) [reply COLLABORATOR_GUIDE.md]: I understood the initial copy but I can appreciate there may be some confusion. Perhaps we can clarify with something like this:

```md
The `redirects.json` file specifies two types of operations: rewrites and redirects.

- Rewrites: These operations do not alter the original URL but instead render the content from the rewritten URL. It's important to note that the rewritten URL path must be valid and exist on the website.

- Redirects: Redirect operations, on the other hand, change the original URL to a new one. While these new URLs can be either internal (within the website) or external (leading to a different domain), it is more common for redirects to point externally. 
```

**Pull request merged**